bedrock/base/middleware.py

```diff
@@ -10,57 +10,174 @@
 """
 
 import base64
+import contextlib
 import inspect
 import time
-from warnings import warn
+from functools import wraps
 
 from django.conf import settings
 from django.core.exceptions import MiddlewareNotUsed
-from django.http import Http404, HttpResponse
+from django.http import Http404, HttpResponse, HttpResponseRedirect
+from django.middleware.locale import LocaleMiddleware as DjangoLocaleMiddleware
+from django.utils import translation
 from django.utils.deprecation import MiddlewareMixin
+from django.utils.translation import trans_real
 
 from commonware.middleware import FrameOptionsHeader as OldFrameOptionsHeader
 
 from bedrock.base import metrics
-from lib.l10n_utils import translation
+from bedrock.base.i18n import (
+    check_for_bedrock_language,
+    get_language_from_headers,
+    normalize_language,
+    path_needs_lang_code,
+    split_path_and_polish_lang,
+)
+from lib.l10n_utils import is_root_path_with_no_language_clues
 
-from . import urlresolvers
 
+class BedrockLangCodeFixupMiddleware(MiddlewareMixin):
+    """Middleware focused on prepping a viable, Bedrock-compatible language code
+    in the URL, ready for the rest of the i18n logic.
+
+    It:
+
+    1) Normalises language codes that are in the path - eg en-us -> en-US and also
+    goes to a prefix code if we don't have support (eg de-AT -> de)
+
+    2) If no redirect is needed, sets request.locale to be the normalized
+    lang code we've got from the URL
+
+    Querystrings are preserved in GET redirects.
```

Comment:
**nitpick (non-blocking):** This docstring should probably also mention the support for querystring-based language switching